Documentation/Common-Gotchas.md

```diff
@@ -0,0 +1,232 @@
+# Common Gotchas
+
+There are some common errors that arise when trying to use Argo. Many are due to
+the fact that Argo is so fundamentally different from how you might be used to
+decoding JSON, and others are due to limitations of Swift. This document aims
+to address some of those situations and set you on the right track!
+
+## `curry` limitations
+
+It is very important to fully curry your model’s `init` in order to use Argo.
+Because of this, `Argo` recommends using
+[`Curry`](https://github.com/thoughtbot/Curry) which allows you to change your
+standard `init` functions into fully curried functions. Unfortunately, Swift
+can only compile `curry` for up to 20 arguments in a reasonable amount of time.
+That may sound like a lot, but one can still run up against this limitation
+quite quickly, especially when you are not in control of the JSON API you are
+interfacing with.
+
+The solution is to try to find a way to normalize your data into nested models,
+thus reducing the number of arguments needed for `init`. For simplicity, assume
+for a moment that `curry` was only defined for functions of 5 arguments or less,
+and consider the following model:
+
+```swift
+struct User: Decodable {
+  let id: Int
+  let name: String
+  let bio: String
+  let smallAvatar: String
+  let mediumAvatar: String
+  let largeAvatar: String
+
+  static func decode(j: JSON) -> Decoded<User> {
+    return curry(User.init)
+      <^> j <| "id"
+      <*> j <| "name"
+      <*> j <| "bio"
+      <*> j <| "small_avatar"
+      <*> j <| "medium_avatar"
+      <*> j <| "large_avatar"
+  }
+}
+```
+
+With our assumption on `curry`, we would get a Swift compiler error when doing
+`curry(User.init)` because `User.init` takes 6 arguments. Even though the API
+has decided to send us back separate fields for each avatar size, there is a
+very clear normalization we can do in our model:
+
+```swift
+struct User {
+  let id: Int
+  let name: String
+  let bio: String
+  let avatar: Avatar
+
+  struct Avatar {
```

Comment:
I might avoid nesting this to avoid potentially confusing people into thinking it _has_ to be nested.